File: apps/web/src/lib/api/scratchpad.ts
import type { CompactionRun } from "@pathfinder/shared/generated/types/CompactionRun";
import type { Note } from "@pathfinder/shared/generated/types/Note";
import { compactionRunSchema } from "@pathfinder/shared/generated/zod/compactionRunSchema";
import { noteSchema } from "@pathfinder/shared/generated/zod/noteSchema";
import { queryOptions } from "@tanstack/react-query";
import { z } from "zod";

import { requestJson, requestVoid } from "./http";

const noteListSchema = z.array(noteSchema);
const compactionListSchema = z.array(compactionRunSchema);

export type { CompactionRun, Note };

export async function listScratchpadNotes(
  conversationId: string,
): Promise<Note[]> {
  return requestJson(
    noteListSchema,
    `/api/v1/conversations/${conversationId}/scratchpad/notes`,
  );
}

export async function getScratchpadNote(
  conversationId: string,
  noteId: string,
): Promise<Note> {
  return requestJson(
    noteSchema,
    `/api/v1/conversations/${conversationId}/scratchpad/notes/${noteId}`,
  );
}

export async function patchScratchpadNote(
  conversationId: string,
  noteId: string,
  body: { pinned: boolean },
): Promise<Note> {
  return requestJson(
    noteSchema,
    `/api/v1/conversations/${conversationId}/scratchpad/notes/${noteId}`,
    { method: "PATCH", body },
  );
}

export async function deleteScratchpadNote(
  conversationId: string,
  noteId: string,
): Promise<void> {
  await requestVoid(
    `/api/v1/conversations/${conversationId}/scratchpad/notes/${noteId}`,
    { method: "DELETE" },
  );
}

export async function listScratchpadCompactions(
  conversationId: string,
): Promise<CompactionRun[]> {
  return requestJson(
    compactionListSchema,
    `/api/v1/conversations/${conversationId}/scratchpad/compactions`,
  );
}

export function scratchpadNotesOptions(conversationId: string) {
  return queryOptions({
    queryKey: ["conversations", conversationId, "scratchpad", "notes"] as const,
    queryFn: () => listScratchpadNotes(conversationId),
  });
}

export function scratchpadCompactionsOptions(conversationId: string) {
  return queryOptions({
    queryKey: [
      "conversations",
      conversationId,
      "scratchpad",
      "compactions",
    ] as const,
    queryFn: () => listScratchpadCompactions(conversationId),
  });
}

File: apps/api/src/pathfinder/transport/http/routers/conversations/scratchpad.py
"""Scratchpad HTTP routes — list/get/patch(pin)/delete + compaction audit log."""

from __future__ import annotations

from uuid import UUID

from assistant_core.platform.pydantic_base import CamelModel
from fastapi import APIRouter, status
from pydantic import ConfigDict, Field

from pathfinder.domain.scratchpad.models import CompactionRun, Note
from pathfinder.services.conversations.scratchpad_service import ScratchpadService
from pathfinder.transport.http.deps import CurrentUser, DBSession

router = APIRouter(prefix="/api/v1/conversations", tags=["scratchpad"])


class ScratchpadPatchRequest(CamelModel):
    """User-driven PATCH — only the ``pinned`` flag is settable from the UI.

    ``extra="forbid"`` rejects ``title``/``body``/``summary`` edits with 422
    so scratchpad content stays agent-authored.
    """

    model_config = ConfigDict(extra="forbid")

    pinned: bool = Field(description="New pin state.")


@router.get(
    "/{conversation_id}/scratchpad/notes",
    response_model=list[Note],
    summary="List scratchpad notes for the conversation.",
)
async def list_scratchpad_notes(
    conversation_id: UUID,
    session: DBSession,
    user_id: CurrentUser,
) -> list[Note]:
    return await ScratchpadService(session).list_notes(conversation_id, user_id)


@router.get(
    "/{conversation_id}/scratchpad/notes/{note_id}",
    response_model=Note,
    summary="Get a single scratchpad note.",
)
async def get_scratchpad_note(
    conversation_id: UUID,
    note_id: str,
    session: DBSession,
    user_id: CurrentUser,
) -> Note:
    return await ScratchpadService(session).get_note(conversation_id, note_id, user_id)


@router.patch(
    "/{conversation_id}/scratchpad/notes/{note_id}",
    response_model=Note,
    summary="Pin or unpin a scratchpad note.",
)
async def patch_scratchpad_note(
    conversation_id: UUID,
    note_id: str,
    request: ScratchpadPatchRequest,
    session: DBSession,
    user_id: CurrentUser,
) -> Note:
    return await ScratchpadService(session).set_pinned(
        conversation_id,
        note_id,
        pinned=request.pinned,
        user_id=user_id,
    )


@router.delete(
    "/{conversation_id}/scratchpad/notes/{note_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a scratchpad note.",
)
async def delete_scratchpad_note(
    conversation_id: UUID,
    note_id: str,
    session: DBSession,
    user_id: CurrentUser,
) -> None:
    await ScratchpadService(session).delete_note(conversation_id, note_id, user_id)


@router.get(
    "/{conversation_id}/scratchpad/compactions",
    response_model=list[CompactionRun],
    summary="Audit log for scratchpad compaction runs.",
)
async def list_compactions(
    conversation_id: UUID,
    session: DBSession,
    user_id: CurrentUser,
) -> list[CompactionRun]:
    return await ScratchpadService(session).list_compactions(conversation_id, user_id)
